File: plur1bus-hermes/src/plur1bus_hermes/critical_reply_intent.py
"""Parse only unambiguous replies to a PLUR1BUS Critical Push.

This module deliberately has no Hermes hook integration.  A caller must supply
trusted quoted-message text from a registered host API; untrusted text that
merely contains a reference line must never become a review action.
"""

from __future__ import annotations

import re
from typing import Any

MAX_REPLY_LENGTH = 200
_REF_LINE_RE = re.compile(r"^\s*(?:Referenz|Reference)\s*:\s*([0-9a-f]{5,32})\s*$", re.I | re.M)
_HEADLINES = (
    "plur1bus hat eine erinnerung als möglicherweise besonders wichtig erkannt.",
    "plur1bus recognized a memory as possibly especially important.",
)
_REJECT_RE = re.compile(r"\b(ablehn\w*|abgelehnt|verwerf\w*|verworfen|reject\w*|dismiss\w*|declin\w*)\b|\bnicht\s+hervorheb\w*|\bdon'?t\s+highlight\b", re.I)
_ACCEPT_RE = re.compile(r"\b(akzeptier\w*|annehm\w*|angenommen|bestätig\w*|übernehm\w*|übernommen|hervorheb\w*|accept\w*|approv\w*|confirm\w*|highlight\w*)\b", re.I)
_NEGATED_ACCEPT_RE = re.compile(r"\b(nicht|kein\w*|not|don'?t|no)\b[^.!?\n]{0,24}\b(akzeptier|annehm|bestätig|übernehm|accept|approv|confirm)", re.I)
_ALL_RE = re.compile(r"\b(alle|alles|allesamt|sämtliche|beide|all|both|everything)\b", re.I)
_GERMAN_RE = re.compile(r"\b(akzeptier\w*|annehm\w*|angenommen|bestätig\w*|übernehm\w*|übernommen|hervorheb\w*|ablehn\w*|abgelehnt|verwerf\w*|verworfen|alle|alles|sämtliche|beide|bitte|nicht)\b", re.I)


def _squash(value: Any) -> str:
    return " ".join(str(value or "").split()).casefold()


def looks_like_critical_push(text: Any) -> bool:
    """Return whether text carries the fixed PLUR1BUS Critical-Push headline."""
    haystack = _squash(text)
    return bool(haystack and any(headline in haystack for headline in _HEADLINES))


def extract_critical_refs(text: Any) -> list[str]:
    """Extract unique short references from a rendered Critical Push."""
    seen: set[str] = set()
    refs: list[str] = []
    for match in _REF_LINE_RE.finditer(text if isinstance(text, str) else ""):
        ref = match.group(1).lower()
        if ref not in seen:
            seen.add(ref)
            refs.append(ref)
    return refs


def detect_reply_language(body: Any) -> str:
    """Return German when the reply uses German decision language, else English."""
    return "de" if _GERMAN_RE.search(body if isinstance(body, str) else "") else "en"


def parse_critical_reply_intent(body: Any) -> dict[str, Any] | None:
    """Parse an explicit accept/reject decision, returning ``None`` when unsure."""
    text = body.strip() if isinstance(body, str) else ""
    if not text or len(text) > MAX_REPLY_LENGTH:
        return None
    if _REJECT_RE.search(text):
        return {"action": "reject", "all": bool(_ALL_RE.search(text))}
    if _NEGATED_ACCEPT_RE.search(text):
        return None
    if _ACCEPT_RE.search(text):
        return {"action": "accept", "all": bool(_ALL_RE.search(text))}
    return None


def build_critical_reply_command(*, body: Any = None, reply_to_body: Any = None) -> dict[str, Any] | None:
    """Build a review command only for an unambiguous reply to a real push."""
    if not looks_like_critical_push(reply_to_body):
        return None
    refs = extract_critical_refs(reply_to_body)
    intent = parse_critical_reply_intent(body)
    if not refs or intent is None or (len(refs) > 1 and not intent["all"]):
        return None
    return {
        "action": intent["action"],
        "refs": refs,
        "args": f"critical {intent['action']} {' '.join(refs)}",
        "lang": detect_reply_language(body),
    }

File: lib/critical-reply-intent.js
/**
 * lib/critical-reply-intent.js — turn a quoted reply to a Critical Push into
 * a critical-review command.
 *
 * A push names each card by its short reference ("Referenz: 9a019"). When the
 * user quotes such a push and writes "bitte alle akzeptieren" or "alle
 * ablehnen", the host hands the quoted text over as `replyToBody`; this
 * module reads the references out of the quote and the decision out of the
 * reply, nothing else. It never guesses: no references in the quote, no clear
 * verb, or several references without "alle" mean "not for me" and the
 * message goes to the agent as usual.
 */

import { t } from "./i18n.js";

const REF_LINE_RE = /^\s*(?:Referenz|Reference)\s*:\s*([0-9a-f]{5,32})\s*$/gim;
const PUSH_LANGUAGES = ["de", "en"];

const squash = (value) => String(value ?? "").replace(/\s+/g, " ").trim().toLowerCase();

/**
 * Is the quoted text one of PLUR1BUS's own pushes? The host names the quoted
 * message's sender only as a display label, so the fixed headline the push
 * always carries is the check: a hand-typed "Referenz: …" line in some other
 * message is not a push and never triggers a decision.
 */
export function looksLikeCriticalPush(text) {
  const haystack = squash(text);
  if (!haystack) return false;
  return PUSH_LANGUAGES.some((lang) => {
    const headline = squash(t("critical.headline", { lang }));
    return headline.length > 0 && haystack.includes(headline);
  });
}
const MAX_REPLY_LENGTH = 200;

// German and English, with the inflections a chat message actually uses.
const REJECT_RE = /\b(ablehn\w*|abgelehnt|verwerf\w*|verworfen|reject\w*|dismiss\w*|declin\w*)\b|\bnicht\s+hervorheb\w*|\bdon'?t\s+highlight\b/i;
const ACCEPT_RE = /\b(akzeptier\w*|annehm\w*|angenommen|bestätig\w*|übernehm\w*|übernommen|hervorheb\w*|accept\w*|approv\w*|confirm\w*|highlight\w*)\b/i;
const NEGATED_ACCEPT_RE = /\b(nicht|kein\w*|not|don'?t|no)\b[^.!?\n]{0,24}\b(akzeptier|annehm|bestätig|übernehm|accept|approv|confirm)/i;
const ALL_RE = /\b(alle|alles|allesamt|sämtliche|beide|all|both|everything)\b/i;
// The reply's own words decide the language of the answer: a German verb or
// "alle"/"bitte" reads as German, otherwise English.
const GERMAN_RE = /\b(akzeptier\w*|annehm\w*|angenommen|bestätig\w*|übernehm\w*|übernommen|hervorheb\w*|ablehn\w*|abgelehnt|verwerf\w*|verworfen|alle|alles|sämtliche|beide|bitte|nicht)\b/i;

/** "de" when the reply uses German words, else "en". */
export function detectReplyLanguage(body) {
  return GERMAN_RE.test(typeof body === "string" ? body : "") ? "de" : "en";
}

/** Short references named in a push, in order, without duplicates. */
export function extractCriticalRefs(text) {
  const refs = [];
  const seen = new Set();
  const source = typeof text === "string" ? text : "";
  for (const match of source.matchAll(REF_LINE_RE)) {
    const ref = match[1].toLowerCase();
    if (seen.has(ref)) continue;
    seen.add(ref);
    refs.push(ref);
  }
  return refs;
}

/**
 * The decision a short reply expresses.
 * @returns {{action: "accept"|"reject", all: boolean} | null}
 */
export function parseCriticalReplyIntent(body) {
  const text = typeof body === "string" ? body.trim() : "";
  if (!text || text.length > MAX_REPLY_LENGTH) return null;
  const all = ALL_RE.test(text);
  if (REJECT_RE.test(text)) return { action: "reject", all };
  if (NEGATED_ACCEPT_RE.test(text)) return null;
  if (ACCEPT_RE.test(text)) return { action: "accept", all };
  return null;
}

/**
 * The critical-review command a quoted reply stands for, or null when the
 * message is not an unambiguous decision about a quoted push.
 * @returns {{action: string, refs: string[], args: string} | null}
 */
export function buildCriticalReplyCommand({ body, replyToBody } = {}) {
  if (!looksLikeCriticalPush(replyToBody)) return null;
  const refs = extractCriticalRefs(replyToBody);
  if (refs.length === 0) return null;
  const intent = parseCriticalReplyIntent(body);
  if (!intent) return null;
  if (!intent.all && refs.length > 1) return null;
  return Object.freeze({
    action: intent.action,
    refs,
    args: `critical ${intent.action} ${refs.join(" ")}`,
    lang: detectReplyLanguage(body),
  });
}
